### PR TITLE
Set town faction

### PR DESCRIPTION
**Pull request by Kert**

### CODE REVIEW

**Comment by jkoppel** [comment]
http://puu.sh/EnXaT/8a57739e32.mp4

Nice work.

I want to see what happens when a town with a building gets changed to one without, and then back.

E.g.: Barbarbarian with Tavern to Necromancer and back
E.g.: Warlock with Black Tower to Sorceress and back.
E.g.: Sorceress with everything to Warlock.

Also, today I learned that the Necromancer Shrine only exists in expansion maps. WTF? I want to see the code that does this.

**Comment by jkoppel** [comment]
(I see the code is very explicit about this, but I still want to see it working.)

File: src/cpp/shared/town/town.cpp [hunk 1]
```diff
@@ -1343,4 +1343,73 @@ int __fastcall CanBuild(town *twn, int building) {
...
+      if(downgradedBldMask & eligibleBuildMask) {
+        this->buildingsBuiltFlags |= downgradedBldMask;
+        this->numCreaturesInDwelling[dwellingIdx] = numCreatures;
+      }
+    }    
+  }
+  
+  // Now we set the offset of graphics for requested faction
+  if(faction == FACTION_CYBORG)
+    faction = (FACTION)6;
```
Comment (jkoppel) [inline src/cpp/shared/town/town.cpp]: This code here changes the interpretation of the "faction" variable for a total of one line.

May as well do something like:

if (faction == FACTION_CYBORG) {
  imgOffset += 6;
} else {
  imgOffset += faction;
}

This preserves the meaning of the "faction" variable.

File: src/cpp/shared/town/town.cpp [hunk 2]
```diff
@@ -1343,4 +1343,73 @@ int __fastcall CanBuild(town *twn, int building) {
...
+      continue;
+
+    // Remove building that can't be built
+    this->buildingsBuiltFlags &= ~(bldMask);
+
+    // Downgrade buildings if upgraded ones don't exist in new faction
+    int dwellingIdx = bldIdx - 19;
+    int numCreatures = this->numCreaturesInDwelling[dwellingIdx];
+    if(bldIdx == BUILDING_UPGRADE_5B) {
+      this->numCreaturesInDwelling[dwellingIdx] = 0;
```
Comment (jkoppel) [inline src/cpp/shared/town/town.cpp]: So Cyclops an change into Black Dragons, but Black Dragons change into nothing?

Comment (Kert) [reply src/cpp/shared/town/town.cpp]: No. Cyclops will change to Green Dragons. Black Dragons will change to Cyclops
The algorithm will never upgrade creatures, only downgrade

**Comment by Kert** [comment]
> Barbarbarian with Tavern to Necromancer and back
Tavern is not built anymore
> Warlock with Black Tower to Sorceress and back.
Black Tower is not built anymore
> Sorceress with everything to Warlock.
Same logic as for the others above

**Comment by Kert** [comment]
> (I see the code is very explicit about this, but I still want to see it working.)

What do you want to see?

File: src/cpp/shared/town/town.cpp (hunk 1 above)
Comment (Kert) [reply src/cpp/shared/town/town.cpp]: Done

**Comment by jkoppel** [comment]
Nice work!

**Pull request merged**